[Casielim] iP

Casielim opened the pull request:
# 🚀 **Vera AI 🤖 ** 💯 

> "No matter how expert you may be, well-designed checklists can improve outcomes." -- Steven Levitt

Vera is a **text-based chatbot** that helps you **manage tasks effortlessly**. It supports **adding, deleting, marking, unmarking, and searching** for tasks while ensuring **persistent storage**.

## 🌟 **Features**
- 📝 **Task Management**
  - Add different types of tasks: **`Todo`**, **`Deadline`**, **`Event`**
  - Delete tasks from the task list
  - Mark and unmark tasks as completed
- 🔍 **Search Functionality**
  - Users can find tasks by searching with a keyword using `find <keyword>`
- 💾 **Persistent Storage**
  - Saves tasks to a local file (`./data/Vera.txt`)
  - Loads saved tasks when the chatbot starts
- ⚠️ **Structured Exception Handling**
  - Handles invalid commands gracefully with `VeraException` 

## 📌 **Installation**
1. **Clone the repository**  
   ```sh
   git clone https://github.com/Casielim/ip.git

2. **Compile the chatbot**  
   ```sh
   javac -d bin -sourcepath src src/main/java/vera/Vera.java
   
3. **Run Vera chatbot**  
   ```sh
   java -cp bin vera.Vera 
  
  ## ✅ Task List
- [x] Implement task management  
- [x] Add search functionality  
- [ ] Implement GUI with JavaFX (coming soon)  
- [ ] Add notifications/reminders (future update)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Casielim committed:
Rename, Greet, Exit

Casielim committed:
Add echo

Casielim committed:
Chatbot customization

Casielim committed:
Add support for list

Casielim committed:
Showing list

Casielim committed:
Add "added:"

Casielim committed:
Add Task.java

Casielim committed:
add mark, unmark, toString feature

Casielim committed:
Apply mark and unmark features

Casielim committed:
Add Deadline.java

Casielim committed:
Add Event.java

Casielim committed:
Add Todo.java

Casielim committed:
Space Adjustment

Casielim committed:
Apply task to chatbot

Casielim committed:
Update input.txt and EXPECTED.TXT

Casielim committed:
Automated Text UI Testing

Casielim committed:
Add VeraException.java

Casielim committed:
Handle errors

Casielim committed:
Add delete feature

Casielim committed:
Handle IndexOutOfRange error

Casielim committed:
Update I/O testing

Casielim committed:
Allow save tasks to file

Casielim committed:
Add exception converting text to task

Casielim committed:
Merge branch 'branch-Level-7'

Casielim committed:
Convert dates and times from String

Casielim committed:
Edit to print diff datetime format

Casielim committed:
Merge branch 'branch-Level-8'

Casielim committed:
Add Ui.java

Casielim committed:
Update input.txt and EXPECTED.TXT

Casielim committed:
Add Storage.java

Casielim committed:
Add Parser.java

Casielim committed:
Add TaskList.java

Casielim committed:
Update input.txt and EXPECTED.TXT

Casielim committed:
Add package

Casielim committed:
Merge branch 'add-gradle-support'

Casielim committed:
Add TodoTest.java

Casielim committed:
TodoTest.java: Update real test code

Casielim committed:
Add DeadlineTest.java

Casielim committed:
TaskList.java: Add getTask method

Casielim committed:
Add TaskListTest.java

Casielim committed:
Edit build.gradle file

Casielim committed:
Add JavaDoc comments

Casielim committed:
Merge branch 'branch-A-JavaDoc'

Casielim committed:
Tweak code to comply with coding standard

Casielim committed:
Merge branch 'branch-A-CodingStandard'

Casielim committed:
Add find feature

Casielim committed:
Merge branch 'branch-Level-9'

lucasisnotcool commented on src/main/java/vera/core/Parser.java:
```diff
@@ -0,0 +1,73 @@
+     * @return a Task object.
+     * @throws IllegalArgumentException If the input format is incorrect or the task type is unrecognised.
+     */
+    public static Task convertTextToTask(String taskText) throws IllegalArgumentException {
```
perhaps this method should be broken up into smaller chunks? 

From Code Quality [Avoid Long Methods](https://nus-cs2103-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#code-quality)
```
Avoid long methods as they often contain more information than what the reader can process at a time. 
Consider if shortening is possible when a method goes beyond 30 LoC. 
The bigger the haystack, the harder it is to find a needle.
```

lucasisnotcool commented on src/main/java/vera/tasks/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public String toString() {
+        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("MMM dd yyyy hhmma");
+        return "[D]" + super.toString() + "(by: " + by.format(dtf) + ")";
```
Perhaps 'D' can be placed as a private static final constant? 

Per Code Quality [Avoid Magic Numbers](https://nus-cs2103-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers):
`Avoid any magic literals in general, not just magic numbers.`

lucasisnotcool commented on src/main/java/vera/tasks/Event.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public String toString() {
+        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("MMM dd yyyy hhmma");
+        return String.format("[E]%s(from: %s to: %s)", super.toString(), from.format(dtf), to.format(dtf));
```
notice that this toString() method looks very different from deadline's implementation. Maybe they can be made to use similar implementations? 😮 

Per Code Quality [Do Not 'Trip Up' Reader](https://nus-cs2103-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#do-not-trip-up-reader)
`similar things that look different`

lucasisnotcool commented on src/main/java/vera/tasks/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+     * @param keyWord string use to search in a list of tasks.
+     */
+    public void findTask(String keyWord) {
+        List<Task> foundedTaskList = new ArrayList<>();
```
Perhaps a better name can be used here? 🤔

lucasisnotcool commented on src/main/java/vera/tasks/Task.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Marks the task as done and print the done message.
+     */
+    public void markFeature() {
```
Hmm I'm not sure what this is supposed to do. It seems like you have a method named markDone but this is different from that? Because markFeature sounds like both noun and verb, perhaps a better explainer name would help?

Per Code Quality [Use Name to Explain](https://nus-cs2103-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-name-to-explain)

lucasisnotcool reviewed (commented):
LGTM. Some minor changes suggested that do not affect functionality 😄

GabriellaGloria commented on src/main/java/vera/Vera.java:
```diff
@@ -0,0 +1,68 @@
+import vera.core.VeraException;
+import vera.tasks.TaskList;
+
+public class Vera{
```
Perhaps add Javadoc comments for the class and its public methods 👍

GabriellaGloria commented on src/main/java/vera/Vera.java:
```diff
@@ -0,0 +1,68 @@
+                list.addTask(cmd);
+            }
+        } catch (NumberFormatException e) {
+            ui.showError(e.getMessage() + " use only index");
```
This error message could be clearer. Consider:
`ui.showError("Invalid input: Please use only numerical indices.");`

GabriellaGloria commented on src/main/java/vera/tasks/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+     * @param s An user input String.
+     * @throws VeraException if the description is missing or the task string format is incorrect.
+     */
+    public void addTask(String s) throws VeraException {
```
Perhaps break this function to smaller parts to avoid it being too long

GabriellaGloria commented on src/main/java/vera/core/Ui.java:
```diff
@@ -0,0 +1,66 @@
+public class Ui {
+    public static final String LINE = "  ________________________________________________"
+            + "________________________";
+    private String greetings = "  Hello! I'm Vera\n  What can I do for you?";
```
Perhaps use constant variables
```java
private static final String GREETING_MESSAGE = "  Hello! I'm Vera\n  What can I do for you?";
private static final String EXIT_MESSAGE = "  Bye. Hope to see you again soon!";`
```

GabriellaGloria reviewed (commented):
LGTM Just a little fix more!

Casielim committed:
TaskList.java: Fix index and double line

Casielim committed:
Ui.java: Modify "Error" to "OOps"

Casielim committed:
Vera.java: Add to catch Exception

Casielim committed:
Implement basic JavaFX

Casielim committed:
Modify naming

Casielim committed:
Modify to return String response for JavaFX UI

Casielim committed:
Create ui package

Casielim committed:
MainWindow.java: Allow app to terminate with "bye"

Casielim committed:
Add background photo

Casielim committed:
Merge branch 'branch-Level-10'

Casielim committed:
Vera.java: Add JavaDoc

Casielim committed:
Task.java: Remove duplicate method

Casielim committed:
Move print statement to UI

Casielim committed:
Allow multiple keywords searching

Casielim committed:
Merge branch 'branch-A-Varargs'

Casielim committed:
Fix checkstyle violation

Casielim committed:
Merge branch 'branch-A-CheckStyle'

Casielim committed:
Vera.java: Add assertions index > 0

Casielim committed:
Parser.java: Reduce code indentation

convertTextToTask method has more than 3 levels of indentation,
which makes the code harder to read and maintain.

The deeper the nesting, the harder it is for the reader to keep track
of the logic. By reducing the number of indentation levels, the logic
becomes clearer and easy to follow.

Refactor the method by initializing a variable and move the try catch
to the higher level improve error handling without additional nesting.
Create a new loadIdDone method to keeps the switch block concise
and focused only on task creation.

Let's decrease the number of levels of indentation to increase code
readability and clarity.

Casielim committed:
Storage.java: Restructure code and throw exception

Restructure the code in loadFileContent method to simplifies the flow
and focuses on processing the lines.

In loadFileContent method, the FileNotFoundException is handled right
at the point where it can be thrown (new Scanner(file)), making it
easier to trace the cause of the error.

Throw exception to signal that there might be an unexpected error
occurred during the program execution.

The code that called the function can then "catch" this exception
using a try-catch block, allowing it to take appropriate actions

Let's
- throws exception for other method to handle
- restructure the code to increase readability

Casielim committed:
TaskList.java: Revise code

Long method often contains more information than what the reader can
process at a time. Shortening the code to smoothen the flow and
increase the quality of code.

Avoid repetitive code to reduce the LoC of addTask method and increase
the readability of the code.

Extract a new method to reduce the level of indentation in findTask
method. The for loop in findTask no longer has nested loops, making it
easier to read. Each method does one thing, improving clarity and
maintainability.

Let's
- avoid long method
- reduce the indentation level of code

Casielim committed:
Vera.java: Catch and handle errors

Casielim committed:
Merge pull request #1 from Casielim/branch-A-Assertions

Vera.java: Add assertions index > 0

Casielim committed:
Merge branch 'master' of https://github.com/Casielim/ip

Casielim committed:
Merge branch 'master' into branch-A-CodeQuality

Casielim committed:
Merge pull request #2 from Casielim/branch-A-CodeQuality

Branch a code quality

Casielim committed:
Add snooze feature

Casielim committed:
Merge branch 'branch-B-Snooze'

Casielim committed:
Add Command.java

Casielim committed:
Change color for Error message for a better GUI

Casielim committed:
Merge branch 'branch-A-BetterGui'

Casielim committed:
Modify Javadoc comments

Casielim committed:
Event.java: Check start time is before end time

Casielim committed:
Reply "Oops" when catching an error

Casielim committed:
Throw VeraException when catching ArrayIndexOutOfBoundsException

Casielim committed:
Handle extra spaces in command

Casielim committed:
Merge branch 'branch-A-MoreErrorHandling'

Casielim committed:
Modify code to follow java coding standard

Casielim committed:
Refactor code structure

Casielim committed:
Change the title to "Vera"

Casielim committed:
Add Ui.png to docs

Casielim committed:
Add README.md

Casielim committed:
Update README.md

Casielim committed:
Update README.md
